File: javascript/alphametics/alphametics.js
let puzzleSolution;

export const solve = (puzzle) => {

  puzzleSolution = {};

  let [operation, result] = puzzle.split(' == ');
  let terms = operation.split(' + ');

  let letters = puzzle.split('').filter(letter => /[A-Z]/g.test(letter));
  letters = Array.from(new Set(letters));

  let numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
  let solution = {};

  solvePuzzle(terms, result, letters, numbers, solution);

  return Object.keys(puzzleSolution).length > 0 ? puzzleSolution : null;

};

// mapeia a solução para os termos dados e 
// testa se essa é uma solução válida
const validSolution = (terms, result, solution) => {

  let t = terms.map(term => term.split('').map(l => solution[l]).join(''));
  let leadingZerosInTerms = t.some(term => term.split('')[0] === '0'); // Testa para 'leading zeros'
  let newTerms = t.map(term => parseInt(term));


  let r = result.split('').map(l => solution[l]).join('');
  let leadingZerosInResult = r.split('')[0] === '0';  // Testa para 'leading zeros'
  let newResult = parseInt(r);


  return newTerms.reduce((acc, i) => i + acc, 0) === newResult 
    && !leadingZerosInTerms && !leadingZerosInResult;
};


// gera todas as soluções possíveis com backtracking
const solvePuzzle = (terms, result, remainingLetters, remainingNumbers, currentSolution) => {

  // --- Caso base --- 
  // Qto foram colocados candidatos para 
  // todas as letras do quebra-cabeças
  // testamos se essa é uma solução válida
  if (remainingLetters.length === 0) {

    if (validSolution(terms, result, currentSolution)) {
      puzzleSolution = JSON.parse(JSON.stringify(currentSolution));
      return true;
    }

    return false;
  }

  // --- Caso recursivo ---
  // Para cada uma das letras ainda não 
  // testadas procura-se um número candidato
  // com backtracking
  let letter = remainingLetters.shift();

  for (let i = 0; i < remainingNumbers.length; i++) {

    let [number] = remainingNumbers.splice(i, 1);
    let newSolution = JSON.parse(JSON.stringify(currentSolution));
    newSolution[letter] = number;

    if (solvePuzzle(terms, result, remainingLetters, remainingNumbers, newSolution)) {
      currentSolution = JSON.parse(JSON.stringify(newSolution));
      return true;
    }

    remainingNumbers.splice(i, 0, number);

  }

  remainingLetters.unshift(letter);

}; 

File: java/alphametics/src/main/java/Alphametics.java
import java.util.*;
import java.util.stream.Collectors;

public class Alphametics {

    private final Puzzle puzzle;
    private Map<Character, Integer> solution;
    private final List<Character> letters;

    public Alphametics(String puzzle) {

        String[] equation = puzzle.split(" == ");
        this.puzzle = new Puzzle(Arrays.asList(equation[0].split(" \\+ ")), equation[1]);
        this.solution = new HashMap<>();
        this.letters = puzzle.replaceAll("[^A-Z]", "").chars()
                .mapToObj(c -> (char) c)
                .distinct()
                .collect(Collectors.toList());

        System.out.println(this);
    }

    public Map<Character, Integer> solve() throws UnsolvablePuzzleException {

        List<Integer> numbers = new ArrayList<>(List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 0));

        if (solvePuzzle(new HashMap<>(), numbers)) {
            return this.solution;
        } else {
            throw new UnsolvablePuzzleException();
        }

    }

    private boolean solvePuzzle(Map<Character, Integer> currentSolution,
                                List<Integer> remainingNumbers) {

        if (this.letters.isEmpty()) {
            if (validSolution(this.puzzle, currentSolution)) {
                this.solution = new HashMap<>(currentSolution);
                return true;
            }

            return false;
        }

        Map<Character, Integer> newSolution = new HashMap<>(currentSolution);
        char currentLetter = this.letters.remove(0);
        int size = remainingNumbers.size();
        for (int i = 0; i < size; i++) {

            int currentNumber = remainingNumbers.remove(i);
            newSolution.put(currentLetter, currentNumber);

            if (solvePuzzle(newSolution, remainingNumbers)) {
                return true;
            } else {
                remainingNumbers.add(i, currentNumber);
                newSolution.remove(currentLetter);
            }

        }

        this.letters.add(0, currentLetter);

        return false;
    }

    private boolean validSolution(Puzzle puzzle, Map<Character, Integer> solution) {

        List<String> lS = puzzle.operands().stream()
                .map(number -> this.wordToNumber(number, solution)).toList();

        if (lS.stream().anyMatch(number -> number.charAt(0) == '0')) { // Testa para leading 0
            return false;
        }

        long leftSide = lS.stream()
                .map(Long::parseLong)
                .mapToLong(Long::longValue)
                .sum();

        String rS = this.wordToNumber(puzzle.result(), solution);
        if (rS.charAt(0) == '0') { // Testa para leading 0
            return false;
        }

        long rightSize = Long.parseLong(rS);

        return leftSide == rightSize;
    }

    private String wordToNumber(String word, Map<Character, Integer> mapping) {
        return word.chars()
                .mapToObj(i -> (char) i)
                .map(c -> mapping.get(c).toString())
                .collect(Collectors.joining());
    }

}